cpp/src/parquet/arrow/path_internal.cc

```diff
@@ -342,7 +341,9 @@ class ListPathNode {
     *child_range = selector_.GetRange(range->start);
     while (child_range->Empty() && !range->Empty()) {
       ++range->start;
-      *child_range = selector_.GetRange(range->start);
+      if (!range->Empty()) {
+        *child_range = selector_.GetRange(range->start);
+      }
     }
```

Comment:
Would be nice to clean up the repetition of conditionals here, for example:
```c++
int64_t empty_elements = 0;
do {
  *child_range = selector_.GetRange(range->start++);
  if (!child_range->Empty()) {
    break;
  }
  ++empty_elements;
} while (!range->Empty());
```

Reply:
done.  needed slightly different code.  range->start is only incremented if child range is empty.